text/0000-impl-specialization.md

```diff
@@ -0,0 +1,1652 @@
+}
+
+impl Add<Complex> for Complex {
+    type Output = Complex;
+    fn add(self, rhs: Complex) {
+        // ...
+    }
+    // no fn add_assign necessary
+}
+```
+
+A particularly nice case of refined defaults comes from trait hierarchies: you
+can sometimes use methods from subtraits to improve default supertrait
+methods. For example, consider the relationship between `size_hint` and
+`ExactSizeIterator`:
+
+```rust
+partial impl<T> Iterator for T where T: ExactSizeIterator {
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.len(), Some(self.len()))
+    }
+}
+```
+
+As we'll see later, the design of this RFC makes it possible to "lock down" such
+method impls (by not using the `default` qualifier), preventing any further
+refinement (akin to Java's `final` keyword); that in turn makes it possible to
+statically enforce the contract that is supposed to connect the `len` and
+`size_hint` methods. (Of course, we can't make *that* particular change, since
+the relevant APIs are already stable.)
+
+## Supporting efficient inheritance
+
+Finally, specialization can be seen as a form of inheritance, since methods
+defined within a blanket impl can be overridden in a fine-grained way by a more
+specialized impl. As we will see, this analogy is a useful guide to the design
+of specialization. But it is more than that: the specialization design proposed
+here is specifically tailored to support "efficient inheritance" schemes (like
+those discussed
+[here](https://internals.rust-lang.org/t/summary-of-efficient-inheritance-rfcs/494))
+without adding an entirely separate inheritance mechanism.
+
+The key insight supporting this design is that virtual method definitions in
+languages like C++ and Java actually encompass two distinct mechanisms: virtual
+dispatch (also known as "late binding") and implementation inheritance. These
+two mechanisms can be separated and addressed independently; this RFC
+encompasses an "implementation inheritance" mechanism distinct from virtual
+dispatch, and useful in a number of other circumstances. But it can be combined
+nicely with an orthogonal mechanism for virtual dispatch to give a complete
+story for the "efficient inheritance" goal that many previous RFCs targeted.
+
+The author is preparing a companion RFC showing how this can be done with a
+relatively small further extension to the language. But it should be said that
+the design in *this* RFC is fully motivated independently of its companion RFC.
+
+# Detailed design
+
+There's a fair amount of material to cover, so we'll start with a basic overview
+of the design in intuitive terms, and then look more formally at a specification.
+
+At the simplest level, specialization is about allowing overlap between impl
+blocks, so long as there is always an unambiguous "winner" for any type falling
+into the overlap. For example:
+
+```rust
+impl<T> Debug for T where T: Display {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```

Comment:
The text hasn't motivated `default` at this point. The example is of what the proposal is trying to allow "at the simplest level"—i.e. before the complexity of needing a `default` keyword is added—not of what the proposed syntax will ultimately look like once this complexity is taken into account. I thought this was clear, but maybe it could be clarified.

It might be a good idea to limit the appearance of motivating examples which don't follow the proposed syntax. One way would just be to add text saying "ignore the `default` keyword for now, it'll be motivated later". This would be unfortunate though, since I liked the style of exposition used here, and adding these caveats would diminish it somewhat.